app/models/run.rb

```diff
@@ -164,6 +164,26 @@ def publish_aging
     publish_age_every(1.day, 30)
   end
 
+  # Calculate the various statistical information about each segments history once in the database for the whole run
+  # instead of individually for each segment (N queries)
+  # Returns an array of SegmentHistories with the following attributes:
+  # segment_id, standard deviation, mean, median, and the 10th, 90th, 99th percentiles
+  def segment_history_stats
+    SegmentHistory.joins(segment: :run)
+                  .where(segment: {runs: {id: id}})
+                  .where.not(realtime_duration_ms: [0, nil])
+                  .group(:segment_id)
+                  .select('
+                    segment_id,
+                    STDDEV_POP(segment_histories.realtime_duration_ms) AS standard_deviation,
+                    AVG(segment_histories.realtime_duration_ms) AS mean,
+                    PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS median,
+                    PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile10,
+                    PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile90,
+                    PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile99
```

Comment:
PERCENTILE_DISC takes the closest element that is actually contained in the set, while PERCENTILE_CONT will calculate it and not necessarily be present in the set.  I thought percentiles were calculated and not strictly a number present but I may have messed that up, it's been a while since I took stats.

Reply:
Any specific reason for calculating 50th percentile discretely but 10th, 90th, and 99th continuously?

Reply:
There is no built in median function, so the workaround is 50th percentile in discrete mode to calculate it.